### PR TITLE
SOLR-14354: HttpShardHandler send requests in async

### PR DESCRIPTION
**Pull request by CaoManhDat**
# Description
Sending request in an async way inside HttpShardHandler

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java [hunk 1]
```diff
@@ -136,6 +137,91 @@ public boolean equals(Object obj) {
     }
   }
 
+  protected static class ServerIterator {
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:140]: I do not see this class very relates to `LBHttp2SolrClient` only and most of the code in LBSolrClient.req() can be replaced by this class.

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 2]
```diff
@@ -64,18 +62,23 @@
    * by the RealtimeGet handler, since other types of replicas shouldn't respond to RTG requests
    */
   public static String ONLY_NRT_REPLICAS = "distribOnlyRealtime";
+  private static final ShardResponse END_QUEUE = new ShardResponse();
 
   private HttpShardHandlerFactory httpShardHandlerFactory;
-  private CompletionService<ShardResponse> completionService;
-  private Set<Future<ShardResponse>> pending;
+  private LinkedList<Cancellable> requests;
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:68]: I think most of the time, max size of the list will equals to number of shards, so I think it won't be very different here but I can change it back to ArrayList.

> I also don't see us removing requests anywhere?
That is right, we need to remove elements from this list on the corresponding result arrived.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 3]
```diff
@@ -359,65 +366,95 @@ public void send(OutStream outStream, SolrRequest req, String collection) throws
...
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (RuntimeException e) {
+      throw e;
+    } catch (Throwable e) {
+      throw createException(req, e);
+    }
+  }
+
+  private SolrServerException createException(Request req, Throwable throwable) {
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:438]: Yeah, the current error handling in SolrClient is quite complex and variety. That is what I can come up at the moment.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 4]
```diff
@@ -359,65 +366,95 @@ public void send(OutStream outStream, SolrRequest req, String collection) throws
...
+      InputStream is = listener.getInputStream();
+      assert ObjectReleaseTracker.track(is);
+      return processErrorsAndResponse(response, parser, is, getEncoding(response), isV2ApiRequest(solrRequest));
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (RuntimeException e) {
+      throw e;
+    } catch (Throwable e) {
+      throw createException(req, e);
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:434]: Right, I think it is worth to seperate the exception handling logic of asyncReq() and req().

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 5]
```diff
@@ -243,12 +233,13 @@ public ShardResponse takeCompletedOrError() {
   }
 
   private ShardResponse take(boolean bailOnError) {
+    try {
+      while (pending.get() > 0) {
+        ShardResponse rsp = responses.take();
+        if (rsp == END_QUEUE)
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:239]: So all usage of HttpShardHandler is single-thread. Submit tasks, takeResponse then cancelAll() if there are anyerror. Therefore even END_QUEUE is not needed here. I may add some comments to indicate that HttpShardHandler is not thread-safe.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java [hunk 6]
```diff
@@ -66,4 +80,132 @@ public LBHttp2SolrClient(Http2SolrClient httpClient, String... baseSolrUrls) {
   protected SolrClient getClient(String baseUrl) {
     return httpClient;
   }
+
+  public interface OnComplete {
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java:84]: Hmm, so the to OnComplete different in the parameter of `onSuccess` one is 
* NamedList
* another is LBSolrClient.Rsp
Should we making them as template then move it to util package of solrj?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java [hunk 7]
```diff
@@ -359,65 +366,95 @@ public void send(OutStream outStream, SolrRequest req, String collection) throws
...
-    Request req = makeRequest(solrRequest, collection);
+  private static final Exception CANCELLED_EXCEPTION = new Exception();
+
+  public Cancellable asyncRequest(SolrRequest solrRequest, String collection, OnComplete onComplete) {
+    Request req;
+    try {
+      req = makeRequest(solrRequest, collection);
+    } catch (SolrServerException | IOException e) {
+      onComplete.onFailure(e);
+      return () -> {};
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:377]: +1

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (madrob) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java:84]: Yea, having a generic interface sounds good.

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 8]
```diff
@@ -54,6 +53,10 @@
 import org.apache.solr.util.tracing.GlobalTracer;
 import org.apache.solr.util.tracing.SolrRequestCarrier;
 
+/**
+ * Submit requests in async manner.
+ * This class is not thread-safe so all methods should be called in a same thread.
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:58]: Use `@SolrSingleThreaded`

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 9]
```diff
@@ -259,12 +261,10 @@ private ShardResponse take(boolean bailOnError) {
 
   @Override
   public void cancelAll() {
-    for (Cancellable cancellable : requests) {
+    for (Cancellable cancellable : responseCancellableMap.values()) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:264]: And clear the map?

Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:264]: +1

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java (hunk 8 above)
Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:58]: +1

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 10]
```diff
@@ -130,77 +134,64 @@ public void submit(final ShardRequest sreq, final String shard, final Modifiable
...
+    }
+    srsp.setShardRequest(sreq);
+    srsp.setShard(shard);
+    SimpleSolrResponse ssr = new SimpleSolrResponse();
+    srsp.setSolrResponse(ssr);
+
+    pending.incrementAndGet();
+    // if there are no shards available for a slice, urls.size()==0
+    if (urls.size() == 0) {
+      // TODO: what's the right error code here? We should use the same thing when
```
Comment (anshumg) [inline solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:156]: Service unavailable error code seems reasonable to me here. Do you have something else you were debating to switch this out to?

Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java:156]: I do not, just copied and pasted from the old code.

File: solr/solrj/src/java/org/apache/solr/client/solrj/util/OnComplete.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.client.solrj.util;
+
+/**
+ * Listener for async requests
+ */
+public interface OnComplete<T> {
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/solrj/src/java/org/apache/solr/client/solrj/util/OnComplete.java]: changed, thanks!

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 12]
```diff
@@ -130,77 +134,64 @@ public void submit(final ShardRequest sreq, final String shard, final Modifiable
...
+      responses.add(srsp);
+      return;
+    }
 
-      try {
-        params.remove(CommonParams.WT); // use default (currently javabin)
-        params.remove(CommonParams.VERSION);
+    // all variables that set inside this listener must be at least volatile
+    responseCancellableMap.put(srsp, this.lbClient.asyncReq(lbReq, new OnComplete<>() {
+      long startTime = System.nanoTime();
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java]: right, thank you Shalin.

File: solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java [hunk 13]
```diff
@@ -66,16 +66,20 @@
   public static String ONLY_NRT_REPLICAS = "distribOnlyRealtime";
 
   private HttpShardHandlerFactory httpShardHandlerFactory;
-  private CompletionService<ShardResponse> completionService;
-  private Set<Future<ShardResponse>> pending;
+  private Map<ShardResponse, Cancellable> responseCancellableMap;
+  private BlockingQueue<ShardResponse> responses;
+  private AtomicInteger pending;
   private Map<String, List<String>> shardToURLs;
   private Http2SolrClient httpClient;
```
[the remark this answers is not in the document]
Comment (CaoManhDat) [reply solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java]: right, I removed that in this commit https://github.com/apache/lucene-solr/pull/1470/commits/0a02baad8587ff6c4f6fce10f06aeeaf88dd8fb2.

**Comment by CaoManhDat** [comment]
Note that this PR including changes for removing the ability to use HttpShardHandler with a custom HttpClient. My reasons for doing that is
* In current code, we use HttpShardHandlerFactory.LBClient (which can't be changed) when requests to multiple replicas (replicas are interchangeable to other, i.e for query cases), for query to single replica we will use provided HttpClient which can be changed by calling `HttpShardHandlerFactory.getShardHandler(HttpClient)`. But when people look into `getShardHandler(HttpClient)` method they will assume that all requests (single or multiple replicas) will use the provided client. 
* The client that provided to `getShardHandler` is all places is updateShardHandler.defaultClient. That does not make sense since in the ref-guide mentioned that
> **shardHandlerFactory**
> socketTimeout
The read timeout for intra-cluster query and administrative requests. The default is the same as the distribUpdateSoTimeout specified in the <solrcloud> section.
therefore the client of httpShardHandlerFactory must be used instead of `updateShardHandler.defaultClient`
* The changes also lead to less connections usage by switching from Http/1.1 client to Http/2

**Pull request merged**